lib/css-converter.js

```diff
@@ -0,0 +1,307 @@
+ * @property {?string} tagName The tag name (aka: type selector https://developer.mozilla.org/en-US/docs/Web/CSS/Type_selectors)
+ * @property {?string[]} classNames An array of CSS class names
+ * @property {?CssAttr[]} attrs An array of CSS attributes
+ * @property {?CssPseudo[]} attrs An array of CSS pseudos
+ * @property {?string} id CSS identifier
+ * @property {?CssRule} rule A descendant of this CSS rule
+ */
+
+/**
+ * Convert a CSS rule to a UiSelector
+ * @param {CssRule} cssRule CSS rule definition
+ */
+function parseCssRule (cssRule) {
+  const { nestingOperator } = cssRule;
+  if (nestingOperator && nestingOperator !== ' ') {
+    log.errorAndThrow(`'${nestingOperator}' is not a supported combinator. ` +
+      `Only child combinator (>) and descendant combinator are supported.`);
+  }
+
+  let uiAutomatorSelector = 'new UiSelector()';
+  if (cssRule.tagName && cssRule.tagName !== '*') {
+    let androidClass = [cssRule.tagName];
+    for (const cssClassNames of (cssRule.classNames || [])) {
+      androidClass.push(cssClassNames);
+    }
+    uiAutomatorSelector += `.className("${androidClass.join('.')}")`;
+  } else if (cssRule.classNames) {
+    uiAutomatorSelector += `.classNameMatches("\\.${cssRule.classNames.join('\\.')}")`;
+  }
+  if (cssRule.id) {
+    uiAutomatorSelector += `.resourceId("${prependAndroidId(cssRule.id)}")`;
+  }
+  if (cssRule.attrs) {
+    for (const attr of cssRule.attrs) {
+      uiAutomatorSelector += parseAttr(attr);
+    }
+  }
+  if (cssRule.pseudos) {
+    for (const pseudo of cssRule.pseudos) {
+      uiAutomatorSelector += parsePseudo(pseudo);
+    }
+  }
+  if (cssRule.rule) {
+    uiAutomatorSelector += `.childSelector(${parseCssRule(cssRule.rule)})`;
+  }
+  return uiAutomatorSelector;
+}
+
+/**
+ * @typedef {Object} CssObject
+ * @property {?string} type Type of CSS object. 'rule', 'ruleset' or 'selectors'
+ */
+
+/**
+ * Convert CSS object to UiAutomator2 selector
+ * @param {CssObject} css CSS object
+ * @returns {string} The CSS object parsed as a UiSelector
+ */
+function parseCssObject (css) {
+  switch (css.type) {
+    case 'rule':
+      return parseCssRule(css);
+    case 'ruleSet':
+      return parseCssObject(css.rule);
+    case 'selectors':
+      return css.selectors.map((selector) => parseCssObject(selector)).join('; ');
+
+    default:
+      // This is never reachable, but if it ever is do this.
+      throw new Error(`UiAutomator does not support '${css.type}' css. Only supports 'rule', 'ruleSet', 'selectors' `);
+  }
+}
+
+/**
+ * Convert a CSS selector to a UiAutomator2 selector
+ * @param {string} cssSelector CSS Selector
+ * @returns {string} The CSS selector converted to a UiSelector
+ */
+CssConverter.toUiAutomatorSelector = function toUiAutomatorSelector (cssSelector) {
+  let cssObj;
+  try {
+    cssObj = parser.parse(cssSelector);
+  } catch (e) {
+    log.errorAndThrow(`Could not parse CSS '${cssSelector}'. Reason: '${e}'`);
+  }
+  return parseCssObject(cssObj);
```

Comment:
what if we mix these helper function into cssObj to avoid passing cssObj argument around:

```
function parseRule () {
...}
....

// was function parseCssObject (css)
function parse () {
     switch (this.type) {
     case 'rule':
       return this.parseRule();
...
}

...

cssObj.parse = parse.bind(cssObj);
cssObj.parseRule = parseRule.bind(cssObj);
```

Reply:
I don't think we're going to agree on this one 😄  ... I'd be happy to have a third Appium contributor be a tie-breaker.

Reply:
I don't want to block the PR. You've done a great work and it would be a waste not to have it merged.

@jlipps Do you have any more comments or suggestions on the PR?

Reply:
Thanks @mykola-mokhnach !